File: array/c/17_1.c
/*
 * find the duplicates in O(n) and O(1) space, given 0<=arr[i]<=n-1
 * */
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

void printDuplicates(int *arr, int n)
{
    int i;
    for (i = 0; i < n; ++i)
    {
        // check sign of arr[arr[i]]
        if (arr[abs(arr[i])] >= 0)// positive
            arr[abs(arr[i])] = -arr[abs(arr[i])];
        else
            printf("%d ", abs(arr[i]));// repeated
    }
}

int main()
{
    int n, i;
    scanf("%d", &n);
    int *arr = malloc(sizeof(int)*n);
    for (i = 0; i < n; ++i)
        scanf("%d", &arr[i]);

    printDuplicates(arr, n);

    return 0;
}

File: array/python/17_1.py
def printDup(arr):
    for i in range(0, len(arr)):
        val = abs(arr[i])
        if arr[val] < 0:  # i.e. it's negative
            print val,
        else:  # its +ve
            arr[val] = -1 * arr[val]


if __name__ == '__main__':
    inp = [9, 6, 5, 0, 8, 2, 7, 1, 8, 5, 8, 6]
    printDup(inp)

# time = O(n) space = O(1)
